proposals/nnnn-move-function.md

```diff
@@ -0,0 +1,455 @@
+@_transparent
+@alwaysEmitIntoClient
+func move<T>(_ t: __owned T) -> T {
+  Builtin.move(t)
+}
+```
+
+Builtin.move is a hook in the compiler to force emission of special SIL "move"
+instructions. These move instructions trigger in the SILOptimizer two special
+diagnostic passes that prove that the underlying binding does not have any uses
+that are reachable from the move using a flow sensitive dataflow. Since it is
+flow sensitive, one is able to end the lifetime of a value conditionally:
+
+```
+if (...) {
+  let y = move(x)
+  // I can't use x anymore here!
+} else {
+  // I can still use x here!
+}
+// But I can't use x here.
+```
+
+This works because the diagnostic passes are able to take advantage of
+control-flow information already tracked by the optimizer to identify all places
+where a variable use could possible following passing the variable to as an
+argument to `move()`.
+
+In practice, the way to think about this dataflow is to think about paths
+through the program. Consider our previous example with some annotations:
+
+```
+let x = ...
+// [PATH1][PATH2]
+if (...) {
+  // [PATH1] (if true)
+  let _ = move(x)
+  // I can't use x anymore here!
+} else {
+  // [PATH2] (else)
+  // I can still use x here!
+}
+// [PATH1][PATH2] (continuation)
+// But I can't use x here.
+```
+
+in this example, there are only 2 program paths, the `[PATH1]` that goes through
+the if true scope and into the continuation and `[PATH2]` through the else into
+the continuation. Notice how the move only occurs along `[PATH1]` but that since
+`[PATH1]` goes through the continuation that one can not use x again in the
+continuation despite `[PATH2]` being safe.
+
+If one works with vars, the analysis is exactly the same except that one can
+conditionally re-initialize the var and thus be able to use it in the
+continuation path. Consider the following example:
+
+```
+var x = ...
+// [PATH1][PATH2]
+if ... {
+  // [PATH1] (if true)
+  let _ = move(x)
+  // I can't use x anymore here!
+  useX(x) // !! ERROR! Use after move.
+  x = newValue
+  // But now that I have re-assigned into x a new value, I can use the var
+  // again.
+} else {
+  // [PATH2] (else)
+  // I can still use x here!
+}
+// [PATH1][PATH2] (continuation)
+// Since I reinitialized x along [PATH1] I can reuse the var here.
+```
+
+Notice how in the above, we are able to use `x` both in the true block AND the
+continuation block since over all paths, x now has a valid value.
+
+The value based analysis uses Ownership SSA to determine if values are used
+after the move and handles non-address only lets. The address based analysis is
+an SSA based analysis that determines if any uses of an address are reachable
+from a move. All of these are already in tree and can be used today by invoking
+the stdlib non-API function `_move` on a local let or move. *NOTE* This function
+is always emit into client and transparent so there isn't an ABI impact so it is
```

Comment:
This NOTE is useful to keep, since it explains the library implementation shown at the top of this section.